Env reg variant

memcpy-rand-rand-rand opened the pull request:
### Prerequisites

- [X] I have read and understood the `CONTRIBUTING` guide
- [X] The commit message follows the [conventional commits][cc] guidelines
- [X] Tests for the changes have been added (for bug fixes/features)
- [] Docs have been added / updated (for bug fixes/features)

### Description

refactor(winreg): change getWindowsRegistryKeyValue fn to be more generic, simplify winreg path

<!---

Tips:

If you're not comfortable with working with Git, we're working a guide (https://ohmyposh.dev/docs/contributing_git) to help you out.
Oh My Posh advises GitKraken (https://www.gitkraken.com/invite/nQmDPR9D) as your preferred cross platform Git GUI power tool.

-->

[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary

memcpy-rand-rand-rand committed:
refactor(winreg): make windows registry query function more general

memcpy-rand-rand-rand committed:
refactor(winreg): make getWindowsRegistryKeyValue more generic

add qword type
adjust logic in winreg segment

memcpy-rand-rand-rand committed:
refactor(winreg): test uses updated reg query API

memcpy-rand-rand-rand commented:
Draft PR for discussion.  Added variant-style code to the getWindowsRegistryKeyValue function to make it more useful should others with to use it.

memcpy-rand-rand-rand committed:
fix(winreg): lint gofmt errors

memcpy-rand-rand-rand committed:
Merge branch 'JanDeDobbeleer:main' into env_reg_variant

memcpy-rand-rand-rand committed:
fix(winreg): gofmt lint error

memcpy-rand-rand-rand committed:
docs(winreg): show fallback in use, add support for QWORD

memcpy-rand-rand-rand committed:
Merge branch 'env_reg_variant' of https://github.com/memcpy-rand-rand-rand/oh-my-posh into env_reg_variant

memcpy-rand-rand-rand committed:
refactor(winreg)!: single query path includes key

remove the need for separate path and key parameters.
all cases can be handled with a single path including the key.
may need users to tweak their configs (hence possibly breaking)

memcpy-rand-rand-rand committed:
fix(winreg): refactor bug using wrong path variable for query

memcpy-rand-rand-rand committed:
feat(winreg): add functions for advanced templates

memcpy-rand-rand-rand committed:
fix(winreg): fix markdown lint error

JanDeDobbeleer commented on src/environment.go:
```diff
@@ -56,6 +56,21 @@ type cache interface {
 
+type windowsRegistryValueType int
+
+const (
```
should we move these to `environment_windows` so they don't get added for other platforms?

memcpy-rand-rand-rand replied:
I was wondering about that too - if possible to have the same interface, but on windows the windowsRegistryValueType and interface etc is full featured where on other platforms it's an interface with no info... I'll have a try.

memcpy-rand-rand-rand replied:
ah - problem with that, segment_winreg.go has a dependency on that interface and enum now, so to make it windows-only at an environment level means the segment would then have to have different platform impls to avoid using the interface/enum.

JanDeDobbeleer replied:
Gotcha!

JanDeDobbeleer commented on src/environment_windows.go:
```diff
@@ -105,31 +105,57 @@ func (env *environment) getCachePath() string {
+	//
+	regPathParts := strings.SplitN(path, "\\", 2)
 
 	regRootHKeyHandle := getHKEYHandleFromAbbrString(regPathParts[0])
```
I didn't catch this at first, it assumes there's an element in the array.

memcpy-rand-rand-rand replied:
yes - I'm making the assumption that the split works - will add some validation.

JanDeDobbeleer commented on src/environment_windows.go:
```diff
@@ -105,31 +105,57 @@ func (env *environment) getCachePath() string {
 	}
 
+	// Strip key off the end.
+	lastSlash := strings.LastIndex(regPathParts[1], "\\")
```
same here, we expect as second element, I would validate if we have all entries before continuing.

memcpy-rand-rand-rand replied:
yup - validation needed.

JanDeDobbeleer commented on src/environment_windows.go:
```diff
@@ -105,31 +105,57 @@ func (env *environment) getCachePath() string {
+	regPath := regPathParts[1][0:lastSlash]
+
+	regKeyLogged := regKey
+	if regKeyLogged == "" {
```
purely style, I always use `len(regKeyLogged) == 0`

JanDeDobbeleer commented on src/environment_windows.go:
```diff
@@ -105,31 +105,57 @@ func (env *environment) getCachePath() string {
+
 	// Second part of split is registry path after HK part - needs to be UTF16 to pass to the windows. API
-	regPathUTF16, regPathUTF16ConversionErr := windows.UTF16FromString(regPathParts[1])
+	regPathUTF16, regPathUTF16ConversionErr := windows.UTF16FromString(regPath)
```
pls rename to `err` for consistency, you can reuse it.

JanDeDobbeleer commented on src/environment_windows.go:
```diff
@@ -105,31 +105,57 @@ func (env *environment) getCachePath() string {
+	if regKeyLogged == "" {
+		regKeyLogged = "(Default)"
+	}
+	env.log(Error, "getWindowsRegistryKeyValue", fmt.Sprintf("getWindowsRegistryKeyValue: root:\"%s\", path:\"%s\", key:\"%s\"", regPathParts[0], regPath, regKeyLogged))
```
why is this line here? Seems to always log that.

memcpy-rand-rand-rand replied:
yes - mostly for debug implementation help.  Likely end-users wouldn't need it I guess.

memcpy-rand-rand-rand replied:
ah whoops - I think I see what you mean, it's always logging an error not debug.

JanDeDobbeleer commented on src/segment_winreg.go:
```diff
@@ -62,3 +74,45 @@ func (wr *winreg) templateString(segmentTemplate string) string {
 	return text
 }
+
+func (wr winreg) GetRegistryString(path string) (string, error) {
```
Oh boy do I want generics in go ASAP...

JanDeDobbeleer commented on src/segment_winreg.go:
```diff
@@ -62,3 +74,45 @@ func (wr *winreg) templateString(segmentTemplate string) string {
+func (wr winreg) GetRegistryString(path string) (string, error) {
+	regValue, err := wr.env.getWindowsRegistryKeyValue(path)
+
+	if regValue != nil {
```
Inverting this if allows the send if to be on the same level below.

memcpy-rand-rand-rand replied:
ooh, good spot - yeah, makes it a lot cleaner 👍

JanDeDobbeleer reviewed (changes requested):
A few remarks, nothing major. Pretty cool stuff tbh.

memcpy-rand-rand-rand committed:
feat(winreg): PR feedback fixes 1

drop-through return logic simplifcation
better validation on supplied reg path
use err

memcpy-rand-rand-rand committed:
feat(winreg): PR feedback fixes 2

move value "variant" only needs to exist in windows impl
fix erroneous "error" log - should have been debug

memcpy-rand-rand-rand committed:
fix(winreg): undo variant move to env_windows

memcpy-rand-rand-rand committed:
Merge branch 'JanDeDobbeleer:main' into env_reg_variant

memcpy-rand-rand-rand commented:
@JanDeDobbeleer I branched this and squashed commits to remove the commit lint issue; it has the changes as reviewed in this PR.  Would you prefer I create a PR on the new branch?

JanDeDobbeleer commented:
@memcpy-rand-rand-rand works for me.
